keps/sig-apps/19-Graduate-CronJob-to-Stable/README.md

```diff
@@ -0,0 +1,607 @@
+
+* **What specific metrics should inform a rollback?**
+  Unexpected restarts of kube-controller-manager.
+
+* **Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?**
+  Manual test is planned once the implementation is finished.
+
+* **Is the rollout accompanied by any deprecations and/or removals of features, APIs,
+fields of API types, flags, etc.?**
+  batch/v1beta1.CronJob is already deprecated.
+
+### Monitoring Requirements
+
+* **How can an operator determine if the feature is in use by workloads?**
+  For the new controller part checking `cronjob_job_creation_skew` and `cronjob_controller_rate_limiter_use` metrics, accompanied with a set of queue-related metrics the old controller is not using.  For the CronJob resource GA verifying if CronJob exists in `batch/v1` API group in the cluster.
+
+* **What are the SLIs (Service Level Indicators) an operator can use to determine
+the health of the service?**
+  - [x] Metrics
+    - Metric name: `cronjob_controller_rate_limiter_use`
+    - Metric name: `cronjob_job_creation_skew`
+    - Metric name: `workqueue_depth`
+    - Metric name: `workqueue_retries`
+    - Metric name: `workqueue_adds_total`
+    - Components exposing the metric: `kube-controller-manager`
+
+* **What are the reasonable SLOs (Service Level Objectives) for the above SLIs?**
+  99th percentile of cron_job_creation_skew <= X per cluster-day.
+
+* **Are there any missing metrics that would be useful to have to improve observability
+of this feature?**
+  No.
+
+### Dependencies
+
+* **Does this feature depend on any specific services running in the cluster?**
+  No.
+
+### Scalability
+
+* **Will enabling / using this feature result in any new API calls?**
+  - WATCH cronjobs, which will replace the current cronjobs polling
+  - estimated throughput TBD
+  - originating component(s) kube-controller-manager
+
+* **Will enabling / using this feature result in introducing new API types?**
+  Graduating CronJob API to GA, iow. CronJob in `batch/v1`.
+
+* **Will enabling / using this feature result in any new calls to the cloud
+provider?**
+  No.
+
+* **Will enabling / using this feature result in increasing size or count of
+the existing API objects?**
+  No.
+
+* **Will enabling / using this feature result in increasing time taken by any
+operations covered by [existing SLIs/SLOs]?**
+  No.
+
+* **Will enabling / using this feature result in non-negligible increase of
+resource usage (CPU, RAM, disk, IO, ...) in any components?**
+  No.
+
+### Troubleshooting
+
+The Troubleshooting section currently serves the `Playbook` role. We may consider
+splitting it into a dedicated `Playbook` document (potentially with some monitoring
+details). For now, we leave it here.
+
+* **How does this feature react if the API server and/or etcd is unavailable?**
+  The controller will not be able to work.
+
+* **What are other known failure modes?**
+  None.
```

Comment:
@soltysh this controller counts on the "ttl-after-finished-controller" one to remove old jobs, right?

Reply:
Nope, cronjob controller cleans its own jobs, CronJob has `.spec.successfulJobsHistoryLimit` and `.spec.failedJobsHistoryLimit` which control that bit. ttl-after-finished-controller relies on Job's `.spec.ttlSecondsAfterFinished`